code/ship/ship.cpp

```diff
@@ -14673,6 +14670,55 @@ int ship_navigation_ok_to_warp(ship *sp)
 	return 1;
 }
 
+// wookieejedi
+// checks both the warp flags and ship_engine_ok_to_warp() and ship_navigation_ok_to_warp() --wookieejedi
+// returns 1 if ship is able to warp, otherwise return 0
+int ship_can_warp_full_check(ship* sp)
+{
+	if (!(sp->cannot_warp_flags()) && ship_navigation_ok_to_warp(sp) && ship_engine_ok_to_warp(sp)) {
+		// there are no warp flags that are preventing us from warping and the nav and engines are good
+		// thus warp is allowed
+		return 1;
+	} else {
+		// warp is not allowed since one of those conditions was not met
+		return 0;
+	}
+}
+
+// wookieejedi
+// check to see if a ship can depart via bay 
+// takes into account if the ship is in a wing
+// returns 1 if the ship has a bay departure and the mothership is present, 0 otherwise
+int ship_can_bay_depart(ship* sp)
+{
+	// if this ship belongs to a wing, then use the wing departure information
+	int departure_location;
+	int departure_anchor;
+	int departure_path_mask;
+	if (sp->wingnum >= 0)
+	{
+		wing *wingp = &Wings[sp->wingnum];
+		departure_location = wingp->departure_location;
+		departure_anchor = wingp->departure_anchor;
+		departure_path_mask = wingp->departure_path_mask;
+	} else {
+		departure_location = sp->departure_location;
+		departure_anchor = sp->departure_anchor;
+		departure_path_mask = sp->departure_path_mask;
+	}
+	
+	if ( departure_location == DEPART_AT_DOCK_BAY )
+	{
+		int anchor_shipnum = ship_name_lookup(Parse_names[departure_anchor]);
```

Comment:
The `Assertion` needs to go *here*, i.e. immediately before this line.  This is because it is checking the array access.  It is okay for the `departure_anchor` to be out-of-range in other places because when it isn't used it is -1.

Reply:
Good. :)